### ISSUE TITLE
[Feature] Overlay Mask Support

### ISSUE THREAD

**Issue by WolfieWaffle** [Captain-Chaos/WorldPainter#1]
Basically this would allow you to supply an image overlay (which already works) but also have an option to use that overlay as a mask. White areas would be 100%, black would be 0%, grey would be 50%, etc. You could use this as a mask for, say, terrain types and use a flow map from world machine. This would be super helpful, and would allow you to do things like add sand on erosion spots such as in a flow map. I don't really have much experience with coding, but I believe that this would not be super-hard to implement. I know that I am not the only one who wants this:

https://www.minecraftforum.net/forums/mapping-and-modding/minecraft-tools/1261257-worldpainter-v1-11-1-graphical-interactive-map?page=313#c6271

https://www.reddit.com/r/Worldpainter/comments/2cqrn5/request_use_overlay_as_mask/

https://www.reddit.com/r/Worldpainter/comments/1ycm0h/using_flow_maps_from_world_machine_in/

I hope you consider adding this feature!

**Comment by Captain-Chaos** [comment]
Cool, my first issue om GitHub... :)

**Comment by Captain-Chaos** [comment]
It's a popular request. Unfortunately it would be a lot of work to implement and I still think it would be relatively niche and advanced. So while it's definitely on the list of things to do it's nearer the medium term than the short term...

**Comment by Captain-Chaos** [comment]
In the mean time however you *can* already do this with scripting. Did you know that? I can give you more details if you're interested.

**Comment by WolfieWaffle** [comment]
No I didn't know that, but that sounds interesting. I was actually
interested in scripting, but did not know how to do it.

**Comment by Captain-Chaos** [comment]
It's documented, somewhat, here: http://www.worldpainter.net/trac/wiki/Scripting See how far you get and let me know if you have questions.

**Comment by WolfieWaffle** [comment]
Ok, It's kind of working, but my world is completely filled to the world
limit when I open it. Every pixel is at height 255. Here is my .js file:
http://pastebin.com/XyvuqrR7

I do have a heightmap in the same location or the script and the command
line is in the same current directory.

What am I doing wrong?

**Comment by Captain-Chaos** [comment]
I'm not sure. The script looks good. It assumes and 8-bit grey scale height map. Are you sure your file is 8-bit and grey scale? If you can't figure it out post your image so I can take a look.

**Comment by WolfieWaffle** [comment]
Well, here is my image:

​
I'll try using the 16-bit support.

**Comment by Captain-Chaos** [comment]
I don't see an image. You're better off replying directly on github instead of replying to emails. You can attach images here.

**Comment by WolfieWaffle** [comment]
Oops, here it is:
![testdelete](https://cloud.githubusercontent.com/assets/8770485/8146014/aa55a3d0-11de-11e5-8b73-d4a5350d08c4.png)

**Comment by Captain-Chaos** [comment]
Thx. It's late here, so I'll look into it later!

**Comment by WolfieWaffle** [comment]
Ok, thanks!

**Comment by Captain-Chaos** [comment]
Did you try the 16-bit support? Because I looked into it, and that is indeed your issue. That height map is 16-bit (meaning the values go from 0 through 65535 instead of 0 through 255). If you adjust the script for that it works fine. You need to adjust all the places where values from the height map are specified, so also the terrain mapping:

```javascript
var world = wp.createWorld().fromHeightMap(heightMap)
        .fromLevels(0, 65535).toLevels(0, 255)
        .go();
 
wp.applyHeightMap(heightMap)
        .toWorld(world)
        .applyToTerrain()
        .fromLevels(0, 16384).toTerrain(36) // Beaches
        .fromLevels(16384, 24576).toTerrain(0) // Grass
        .fromLevels(24577, 28672).toTerrain(3) // Permadirt
        .fromLevels(28673, 65535).toTerrain(29) // Rock
        .go();
```

**Comment by WolfieWaffle** [comment]
Ok that worked. Just one more question: is it possible to use scripting to do percentages on layers that are either on or off?

Like if you put the brush at 50% strength, with the cobble terrain, is does not completely cover it - some are 100% grass and some 100% are cobble, distributed evenly creating a noise pattern. Is it possible to do that with the engine, or is it only "100% completely or 0%, completely, no mix."

I hope that made sense :P

**Comment by Captain-Chaos** [comment]
No, that's currently not possible with WorldPainter itself. You can do it by changing the height map to a 1-bit black and white bitmap yourself in a paint program using the dithering method.

**Comment by WolfieWaffle** [comment]
Ah, so do the dithering to the heightmap itself. I would probably want a separate hieghtmap for terrain types and actual height then, or I would have a world with only two possible height values! I'm really starting to like the scripting though, another benefit I found was that I can export worldmachines sections separately and stitch the height maps together, rather than in worldpainter.
